pink/barriers/barrier.py

```diff
@@ -0,0 +1,221 @@
+
+        Args:
+            dim: Dimension of the barrier.
+            gain: barrier gain. Defaults to 1.0.
+            class_k_fn: Extended class K function.
+                Defaults to the identity function.
+            r: Weighting factor for the safe backup policy regularization term.
+                Defaults to 3.0.
+        """
+        self.dim = dim
+        self.gain = (
+            gain if isinstance(gain, np.ndarray) else np.ones(dim) * gain
+        )
+
+        self.class_k_fn = class_k_fn if class_k_fn is not None else lambda x: x
+        self.r = r
+        self.safe_policy = np.zeros(self.dim)
+
+    @abc.abstractmethod
+    def compute_barrier(self, configuration: Configuration) -> np.ndarray:
+        r"""Compute the value of the barrier function.
+
+        The barrier function :math:`\boldsymbol{h}(\boldsymbol{q})`
+        is a vector-valued function that represents the safety constraints.
+        It should be designed such that the set
+        :math:`\{\boldsymbol{q} : \boldsymbol{h}(\boldsymbol{q}) \geq \boldsymbol{0}\}`
+        represents the safe region of the configuration space.
+
+        Args:
+            configuration: Robot configuration :math:`\boldsymbol{q}`.
+
+        Returns:
+            Value of the barrier function :math:`\boldsymbol{h}(\boldsymbol{q})`.
+        """  # noqa: E501
+
+    @abc.abstractmethod
+    def compute_jacobian(self, configuration: Configuration) -> np.ndarray:
+        r"""Compute the Jacobian matrix of the barrier function.
+
+        The Jacobian matrix
+        :math:`\frac{\partial \boldsymbol{h}}{\partial \boldsymbol{q}}(\boldsymbol{q})`
+        of the barrier function with respect to the configuration variables is
+        required for the computation of the barrier condition.
+
+        Args:
+            configuration: Robot configuration :math:`\boldsymbol{q}`.
+
+        Returns:
+            Jacobian matrix
+            :math:`\frac{\partial \boldsymbol{h}}{\partial \boldsymbol{q}}(\boldsymbol{q})`.
+        """  # noqa: E501
+
+    def compute_safe_policy(self, configuration: Configuration) -> np.ndarray:
+        r"""Compute the safe backup control policy.
+
+        The safe backup control policy
+        :math:`\dot{\boldsymbol{q}}_{safe}(\boldsymbol{q})` is a joint
+        velocity vector that can be used as a regularization term in the
+        optimization problem to ensure safety.
+
+        Args:
+            configuration: Robot configuration :math:`\boldsymbol{q}`.
+
+        Returns:
+            Safe backup joint velocities
+                :math:`\dot{\boldsymbol{q}}_{safe}(\boldsymbol{q})`.
+        """
+        return np.zeros(configuration.model.nv)
+
+    def compute_qp_objective(
+        self,
+        configuration: Configuration,
+    ) -> Tuple[np.ndarray, np.ndarray]:
+        r"""Compute the quadratic objective function for the barrier-based QP.
+
+        The quadratic objective function includes a regularization term
```

Comment:
If I understand correctly, this regularization term is only added when $r>0$, and the objective is otherwise zero?

If so, I think we should point it out in the docs.